SwiftCompilerSources/Sources/Optimizer/Utilities/AccessUtils.swift

```diff
@@ -0,0 +1,435 @@
+/// %base2 = alloc_stack $S
+/// %base3 = global_addr @gaddr
+/// %addr1 = struct_element_addr %base1
+/// %access1 = store %v1 to [trivial] %addr1   // accessed address is offset from base
+/// %access2 = store %v2 to [trivial] %base2   // accessed address is base itself
+/// ```
+/// The base address is never inside an access scope.
+struct AccessBase : CustomStringConvertible {
+  let baseAddress: Value
+  let storageKind: AccessStorageKind
+  
+  var description: String {
+    "\(storageKind) - \(baseAddress)"
+  }
+  
+  var referenceWithPathComponent: (ref: Value, comp: (SmallProjectionPath.FieldKind, Int))? {
+    switch storageKind {
+    case .Box, .Tail, .Class:
+      switch baseAddress {
+      case let rea as RefElementAddrInst:
+        return (rea.operand, (.classField, rea.fieldIndex))
+      case let pb as ProjectBoxInst:
+        return (pb.operand, (.classField, pb.fieldIndex))
+      case let rta as RefTailAddrInst:
+        return (rta.operand,(.tailElements, 0))
+      default:
+        fatalError("Inconsistent state")
+      }
+    default:
+      return nil
+    }
+  }
+  
+  var reference: Value? { referenceWithPathComponent?.ref }
+  
+  var isLet: Bool {
+    switch baseAddress {
+    case is RefElementAddrInst:
+      fatalError() // FIXME: add support for looking up types in libswift etc
+    default:
+      fatalError()
+    }
+  }
+  
+  var isLocal: Bool {
+    switch storageKind {
+    case .Box:
+      return baseAddress is AllocBoxInst
+    case .Stack:
+      return true
+    default:
+      return false
+    }
+  }
+  
+  /// Returns `true` if we can reliably compare for equality this `AccessBase`
+  /// with another `AccessBase`
+  var isUniquelyIdentified: Bool {
+    switch storageKind {
+    case .Box:
+      return baseAddress is AllocBoxInst
+    case .Stack, .Global:
+      return true
+    case .Argument:
+      return (baseAddress as! FunctionArgument).isExclusiveIndirectParameter
+    default:
+      return false
+    }
+  }
+  
+  func isDistinct(from other: AccessBase) -> Bool {
+    fatalError()
+  }
+}
+
+/// An `AccessPath` is a pair of a `base: AccessBase` and a `projectionPath: Path`
+/// which denotes the offset of the access from the base in terms of projections.
+struct AccessPath : CustomStringConvertible {
+  var base: AccessBase
+
+  /// address projections only
+  var projectionPath: SmallProjectionPath
+  
+  var description: String {
+    "\(projectionPath): \(base)"
+  }
+  
+  func isDistinct(from: AccessPath) -> Bool {
+    fatalError("TODO")
+  }
+}
+
+/// An `AccessRoot` is the reference (or a value for which the
+/// reference is a stored property) an address originates from.
+struct AccessRoot {
+  var root: Value
+  
+  /// Only valid paths are: `"<sequence of value projections>.<one reference projection>.<sequence of address projections>"`
+  var path: SmallProjectionPath
+}
+
+typealias AccessStorage = (maybeRoot: AccessRoot?, base: AccessBase)
```

Comment:
You can see here that both, `maybeRoot` and `base`, contain a path.
BTW, I really prefer to make this a struct instead of a tuple.

Also, this is a bad name, IMO.
If I understood correctly, the storage of an access is either a class, global variable or stack location.
So let's make `AccessStorage` a protocol to which `AllocRefInst`, `GlobalVariable`, `AllocStackInst`, etc. conform to.

Together with a path, it should be named `AccessPathFromStorage` or `AccessStoragePath`.

Reply:
> You can see here that both, maybeRoot and base, contain a path

Only `maybeRoot` contains a path.

> Also, this is a bad name, IMO.
> If I understood correctly, the storage of an access is either a class, global variable or stack location.
> So let's make AccessStorage a protocol to which AllocRefInst, GlobalVariable, AllocStackInst, etc. conform to.
> Together with a path, it should be named AccessPathFromStorage or AccessStoragePath.

Indeed. In my proposal, it looks like I changed the contents of `AccessRoot` to avoid storing multiple paths, but didn't adjust the name.

Let's refer to `AccessRoot` as `AccessStoragePath` henceforth...

Now that the full projection path is part of `AccessStoragePath`, it is no longer specific to reference storage. This significantly changes the design.

Now, rather than containing a `Value`, `AccessStoragePath` should contain the `AccessStorage`:

```
struct AccessStoragePath {
  var storage: AccessStorage
  var path: SmallProjectionPath // both value and address projections
}
```

> So let's make AccessStorage a protocol to which AllocRefInst, GlobalVariable, AllocStackInst, etc. conform to.

Right! Now we can make `AccessStoragePath` valid for global variables in addition to reference storage.

We *could* derive an `AccessRoot` from an `AccessStoragePath` just by chopping off the address projections (future consideration).

Sometimes, we want to identify the accessed storage and *also* need to recover the base address. This is actually pretty common. We just want some simple type that combines these three things
1. identity of the storage
2. path from that storage to the accessed address
3. the base address

In the proposal that "triple" was called `AccessStorage`, which is a nonsense name. If we just follow the naming convention of MemAccessUtils.h, then it would be `AccessStoragePathWithBase`.

Let's review.

Each struct contains an increasing amount of information, starting from an address value.

Roughly in order they are discovered when walking up:

- Address Value (starting point)

- EnclosingScope (AccessBase or AccessScope)

- AccessBase (base address and kind of storage; identifies a formal access)

- AccessPath (base + address projections)

- AccessStorage (identifies a storage object at the lowest granularity)

- AccessStoragePath (storage + value projections + address projections)

- AccessRoot (TBD: reference-type storage + value projections)

(I might update the proposal tomorrow)